tests/vm_test.go

```diff
@@ -2460,45 +2459,105 @@ status:
+
+			err = virtClient.VirtualMachine(testsuite.GetTestNamespace(vm)).Delete(context.Background(), vm.Name, &metav1.DeleteOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(func(g Gomega) {
+				vm, err = virtClient.VirtualMachine(vm.Namespace).Get(context.Background(), vm.Name, &k8smetav1.GetOptions{})
+				g.Expect(err).ToNot(HaveOccurred())
+				g.Expect(controller.HasFinalizer(vm, v1.VirtualMachineControllerFinalizer)).To(BeFalse())
+				g.Expect(controller.HasFinalizer(vm, customFinalizer)).To(BeTrue())
 			}, 2*time.Minute, 1*time.Second)
 		})
 
-		It("should be removed when the vm is being deleted", func() {
+		It("should be removed when the vm has child resources, such as instance type ControllerRevisions, that have been deleted before the vm - issue #9438", func() {
+			By("creating a VirtualMachineClusterInstancetype")
+			instancetype := &instancetypev1alpha2.VirtualMachineClusterInstancetype{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "instancetype-",
+				},
+				Spec: instancetypev1alpha2.VirtualMachineInstancetypeSpec{
+					CPU: instancetypev1alpha2.CPUInstancetype{
+						Guest: uint32(1),
+					},
+					Memory: instancetypev1alpha2.MemoryInstancetype{
+						Guest: resource.MustParse("64Mi"),
+					},
+				},
+			}
+			instancetype, err = virtClient.VirtualMachineClusterInstancetype().Create(context.Background(), instancetype, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			By("creating a VirtualMachine")
+			vm = &v1.VirtualMachine{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "vm-",
+				},
+				Spec: v1.VirtualMachineSpec{
+					Running: pointer.Bool(false),
+					Instancetype: &v1.InstancetypeMatcher{
+						Name: instancetype.Name,
+					},
+					Template: &v1.VirtualMachineInstanceTemplateSpec{
+						Spec: v1.VirtualMachineInstanceSpec{
+							Domain: v1.DomainSpec{},
+						},
+					},
+				},
+			}
 			vm.Finalizers = append(vm.Finalizers, customFinalizer)
 			vm, err = virtClient.VirtualMachine(testsuite.GetTestNamespace(vm)).Create(context.Background(), vm)
 			Expect(err).ToNot(HaveOccurred())
 
+			By("waiting until the VirtualMachine has the VirtualMachineControllerFinalizer, customFinalizer and revisionName")
 			Eventually(func(g Gomega) {
 				vm, err = virtClient.VirtualMachine(vm.Namespace).Get(context.Background(), vm.Name, &k8smetav1.GetOptions{})
 				g.Expect(err).ToNot(HaveOccurred())
 				g.Expect(controller.HasFinalizer(vm, v1.VirtualMachineControllerFinalizer)).To(BeTrue())
+				g.Expect(controller.HasFinalizer(vm, customFinalizer)).To(BeTrue())
+				g.Expect(vm.Spec.Instancetype.RevisionName).ToNot(BeEmpty())
 			}, 2*time.Minute, 1*time.Second)
 
-			err = virtClient.VirtualMachine(testsuite.GetTestNamespace(vm)).Delete(context.Background(), vm.Name, &metav1.DeleteOptions{})
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Get(context.Background(), vm.Name, &k8smetav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
```

Comment:
nit: `vm` is already updated from previous `Eventually`

Reply:
So I assumed the same but the value of `vm` here is still that returned by the `.Create()` call above without `vm.Spec.Instancetype.RevisionName` populated (the `VM` controller adds this after creation). Taking this additional `Get` out causes the `ControllerRevision` `Delete()` to fail as the name provided is empty.

Reply:
I learned something new. Thanks